### ISSUE TITLE
String mis-decoded as Protobuf

### ISSUE THREAD

**Issue by pawitp** [pawitp/protobuf-decoder#5]
The string "123-456-789" is mistakenly decoded as an inner protobuf message. We should provide a way for users to override the decode format.

**Comment by konsumer** [comment]
Yeh, I found this in my own libs, too (see notes on #81 ) One idea I had on protoquery was a sort of "choices" interface. Like if there was a select-box on every branch to say "this should be parsed as a string" you could get better output, because you can say "this LEN field is really a string, even though it's possible to parse as a sub-message" it also works for `packed repeat` fields (see #2 ) Here is the format I imagine for choices:

```json
{
  "1.2.4": "string"
}
``` essentially a flat JSON of associative-array that has full path and expected type. This way we could load a choices object before parsing, and use it for display, and I think it could also help with generating JSON and even proto SDL (using my rawproto lib.) I was thinking about doing my own seperate UI project that uses these ideas, but I would rather merge with yours, so we can share some of this work.

**Comment by pawitp** [comment]
Yes, having a select dropdown box to select the decoding type is what I imagined when I wrote this issue. It would be smart though and:
1. Only show dropdown if there are multiple ways to decode the field
2. In the dropdown, only valid choices are displayed

The dropdown would also be needed for numbers which can be decoded multiple ways, but we are simply displaying all choices at the moment because of lack of dropdown support.

I'm not sure what you mean by GROUPSTART/GROUPEND.

**Comment by konsumer** [comment]
> I'm not sure what you mean by GROUPSTART/GROUPEND.

Wiretype 3 & 4. [Here](https://protobuf.dev/programming-guides/encoding/#groups) they call it `SGROUP` and `EGROUP`.

These are deprecated in proto3, but still show up a lot in old proto (like when parsing google play or other proto2 traffic.) You can see it in my protoquery unit-tests (`hearthstone` is google play detail record.) The example above is a group that does not parse in yours, but works in mine.

Working with groups is pretty simple, essentially you just grab bytes from between `SGROUP` and `EGROUP` and parse as a repeat message. It's always messages, so you cannot directly have other `LEN` types (packed repeated or strings/bytes) at top, but they can be part of the repeated sub-message. It's the only wire-type that has a start & end byte-marker.

I have decided to work on my own separate rawproto UI tool, since I think we have different ideas about how it should fit together, tooling, and how the actual UI should work, and I have already made libs for doing things like parsing the binary tree initially, outputting JSON or proto SDL, and doing queries, that I am anxious to merge into rawproto, along with my own ideas for a UI (lazy message-branch eval, choices, more field-types, etc.) I wrote [a book](https://www.packtpub.com/product/practical-grpc/9781839211744) that uses some of my libs, and even though it's been a little while (5 years or so) I constantly have ideas for making the tooling better.

Good luck!

**Comment by pawitp** [comment]
Good luck with your project!
